test/object-store/util/sync/sync_test_utils.cpp

```diff
@@ -287,26 +287,34 @@ void wait_for_advance(Realm& realm)
     realm.m_binding_context = nullptr;
 }
 
-void async_open_realm(const Realm::Config& config,
-                      util::UniqueFunction<void(ThreadSafeReference&& ref, std::exception_ptr e)> finish)
+StatusWith<std::shared_ptr<Realm>> async_open_realm(const Realm::Config& config)
 {
-    std::mutex mutex;
-    bool did_finish = false;
     auto task = Realm::get_synchronized_realm(config);
-    ThreadSafeReference tsr;
-    std::exception_ptr err = nullptr;
+    auto pf = util::make_promise_future<ThreadSafeReference>();
     task->start([&](ThreadSafeReference&& ref, std::exception_ptr e) {
-        std::lock_guard lock(mutex);
-        did_finish = true;
-        tsr = std::move(ref);
-        err = e;
-    });
-    util::EventLoop::main().run_until([&] {
-        std::lock_guard lock(mutex);
-        return did_finish;
+        if (e) {
+            try {
+                std::rethrow_exception(e);
+            }
+            catch (...) {
+                pf.promise.set_error(exception_to_status());
+            }
+        }
+        else {
+            pf.promise.emplace_value(std::move(ref));
+        }
     });
-    task->cancel(); // don't run the above notifier again on this session
-    finish(std::move(tsr), err);
+    auto sw = std::move(pf.future).get_no_throw();
```

Comment:
nit: i think this could be 
```
return std::move(pf.future).then([&](ThreadSafeReference&& ref) {
    return Realm::get_shared_realm(std::move(ref));
}).get_no_throw();
```

Reply:
No, that'd open the Realm on the wrong thread. Resolving the TSR needs to happen after the get() so that it happens on the main thread.

Future's API works well for passing around thread-unsafe objects but less so for thread-confined objects.